### 0. Imports

In [69]:
%load_ext autoreload
%autoreload 2

# Data transformation
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Visualizations
# -----------------------------------------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt

# Progress loops
# -----------------------------------------------------------------------
from tqdm import tqdm

# Handle warnings
# -----------------------------------------------------------------------
import warnings
warnings.filterwarnings("once")

# modify system variables
# -----------------------------------------------------------------------
import sys
sys.path.append("..") # append parent folder to path

# import support functions
# -----------------------------------------------------------------------
import src.soporte_eda as se
import src.soporte_preprocesamiento as sp
import src.soporte_clustering as sc


# statistics functions
# -----------------------------------------------------------------------
from scipy.stats import pearsonr, spearmanr, pointbiserialr



The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


# 1. Intro to this notebook and data

## 1.1 Introduction

The goal of this busines case is to 1. identify similar groups of clients according to their buying behaviour 2. identify groups of products according to their profitability. For that, analysis, clustering algorithms and regression models will be used.

The purpose of this notebook is to explore Global Ecommerce's product and customers data in order to clean any possible errors that could impair the extraction of insights through analysis and the quality of the subsequent model development for their brand.

## 1.2 Data import

In [70]:
global_superstore = pd.read_csv("../data/Global_Superstore.csv", encoding="latin1")

# transform column names into lowercase and replace spaces by underscores
global_superstore.columns = [col.lower().replace(" ", "_") for col in global_superstore.columns]

global_superstore.head(3)

row_id        order_id  order_date   ship_date     ship_mode customer_id  \
0   32298  CA-2012-124891  31-07-2012  31-07-2012      Same Day    RH-19495   
1   26341   IN-2013-77878  05-02-2013  07-02-2013  Second Class    JR-16210   
2   25330   IN-2013-71249  17-10-2013  18-10-2013   First Class    CR-12730   

   customer_name    segment           city            state        country  \
0    Rick Hansen   Consumer  New York City         New York  United States   
1  Justin Ritter  Corporate     Wollongong  New South Wales      Australia   
2   Craig Reiter   Consumer       Brisbane       Queensland      Australia   

   postal_code market   region       product_id    category sub-category  \
0      10024.0     US     East  TEC-AC-10003033  Technology  Accessories   
1          NaN   APAC  Oceania  FUR-CH-10003950   Furniture       Chairs   
2          NaN   APAC  Oceania  TEC-PH-10004664  Technology       Phones   

                                        product_name     sales  quantity  \
0  Plantronics CS510 - Over-the-Head monaural Wir...  2309.650         7   
1          Novimex Executive Leather Armchair, Black  3709.395         9   
2                  Nokia Smart Phone, with Caller ID  5175.171         9   

   discount    profit  shipping_cost order_priority  
0       0.0  762.1845         933.57       Critical  
1       0.1 -288.7650         923.63       Critical  
2       0.1  919.9710         915.49         Medium

# 2. Preliminary data analysis and cleaning

For a quick initial exploration:

In [71]:
pd.set_option("display.max_columns",24)
se.exploracion_dataframe(global_superstore)

El número de datos es 51290 y el de columnas es 24

 ..................... 

5 filas aleatorias del dataframe son:


row_id        order_id  order_date   ship_date       ship_mode  \
44646   45406    UP-2012-9910  03-05-2012  07-05-2012  Standard Class   
2690     9907  MX-2011-104794  04-04-2011  08-04-2011  Standard Class   
44920   36134  CA-2014-106831  02-06-2014  04-06-2014     First Class   
37986   33794  CA-2012-122371  26-09-2012  01-10-2012  Standard Class   
1673    23187   IN-2013-44005  20-08-2013  27-08-2013  Standard Class   

      customer_id  customer_name      segment              city  \
44646     RD-9900    Ruben Dartt     Consumer          Donets'k   
2690     BW-11065  Barry Weirich     Consumer           Limeira   
44920    FH-14350    Fred Harton     Consumer            Dublin   
37986    BS-11800  Bryan Spruell  Home Office         San Diego   
1673     CR-12580  Clay Rozendal  Home Office  Ho Chi Minh City   

                  state        country  postal_code market          region  \
44646           Donetsk        Ukraine          NaN   EMEA            EMEA   
2690          São Paulo         Brazil          NaN  LATAM           South   
44920              Ohio  United States      43017.0     US            East   
37986        California  United States      92037.0     US            West   
1673   Ho Chí Minh City        Vietnam          NaN   APAC  Southeast Asia   

             product_id         category sub-category  \
44646  OFF-ADV-10002593  Office Supplies    Fasteners   
2690    TEC-CO-10000388       Technology      Copiers   
44920   OFF-BI-10003429  Office Supplies      Binders   
37986   OFF-EN-10000056  Office Supplies    Envelopes   
1673    TEC-PH-10003613       Technology       Phones   

                                            product_name      sales  quantity  \
44646                         Advantus Clamps, Bulk Pack    19.2900         1   
2690                            Canon Fax Machine, Color  1063.9678         5   
44920  Cardinal HOLDit! Binder Insert Strips,Extra St...     3.7980         2   
37986                        Cameo Buff Policy Envelopes   124.4600         2   
1673                          Apple Signal Booster, VoIP  1361.9304        12   

       discount   profit  shipping_cost order_priority  
44646     0.000   7.7100           1.30         Medium  
2690      0.002  -2.1322         107.66           High  
44920     0.700  -2.6586           1.25       Critical  
37986     0.000  58.4962           2.74         Medium  
1673      0.170  49.0104         150.25         Medium


 ..................... 

Los tipos de las columnas y sus valores únicos son:


tipo_dato  conteo
row_id             int64   51290
order_id          object   25035
order_date        object    1430
ship_date         object    1464
ship_mode         object       4
customer_id       object    1590
customer_name     object     795
segment           object       3
city              object    3636
state             object    1094
country           object     147
postal_code      float64     631
market            object       7
region            object      13
product_id        object   10292
category          object       3
sub-category      object      17
product_name      object    3788
sales            float64   22995
quantity           int64      14
discount         float64      27
profit           float64   24575
shipping_cost    float64   10037
order_priority    object       4


 ..................... 

Los duplicados que tenemos en el conjunto de datos son: 0

 ..................... 

Los nulos que tenemos en el conjunto de datos son:


%_nulos
postal_code  80.51472


 ..................... 

Comprobamos que no haya valores con una sola variable:

 ..................... 

Comprobamos una representación mínima para valores numéricos:
● La variable quantity tiene 14 < 15 valores únicos. Se convierte a objeto.

 ..................... 

Estadísticas descriptivas de las columnas numéricas:


count          mean           std       min           25%  \
row_id         51290.0  25645.500000  14806.291990     1.000  12823.250000   
postal_code     9994.0  55190.379428  32063.693350  1040.000  23223.000000   
sales          51290.0    246.490581    487.565361     0.444     30.758625   
discount       51290.0      0.142908      0.212280     0.000      0.000000   
profit         51290.0     28.610982    174.340972 -6599.978      0.000000   
shipping_cost  51290.0     26.375915     57.296804     0.000      2.610000   

                     50%         75%        max  
row_id         25645.500  38467.7500  51290.000  
postal_code    56430.500  90008.0000  99301.000  
sales             85.053    251.0532  22638.480  
discount           0.000      0.2000      0.850  
profit             9.240     36.8100   8399.976  
shipping_cost      7.790     24.4500    933.570


 ..................... 

Estadísticas descriptivas de las columnas categóricas:


count unique              top   freq
order_id        51290  25035   CA-2014-100111     14
order_date      51290   1430       18-06-2014    135
ship_date       51290   1464       22-11-2014    130
ship_mode       51290      4   Standard Class  30775
customer_id     51290   1590         PO-18850     97
customer_name   51290    795  Muhammed Yedwab    108
segment         51290      3         Consumer  26518
city            51290   3636    New York City    915
state           51290   1094       California   2001
country         51290    147    United States   9994
market          51290      7             APAC  11002
region          51290     13          Central  11117
product_id      51290  10292  OFF-AR-10003651     35
category        51290      3  Office Supplies  31273
sub-category    51290     17          Binders   6152
product_name    51290   3788          Staples    227
quantity        51290     14                2  12748
order_priority  51290      4           Medium  29433


 ..................... 

Los valores que tenemos para las columnas categóricas son: 
La columna ORDER_ID tiene 25035 valores únicos, de los cuales los primeros son:


count  pct
order_id                  
CA-2014-100111     14  0.0
IN-2012-41261      13  0.0
TO-2014-9950       13  0.0
IN-2013-42311      13  0.0
NI-2014-8880       13  0.0

La columna ORDER_DATE tiene 1430 valores únicos, de los cuales los primeros son:


count  pct
order_date            
18-06-2014    135  0.3
18-11-2014    127  0.2
03-09-2014    126  0.2
20-11-2014    118  0.2
29-12-2014    116  0.2

La columna SHIP_DATE tiene 1464 valores únicos, de los cuales los primeros son:


count  pct
ship_date             
22-11-2014    130  0.3
07-09-2014    115  0.2
07-12-2014    101  0.2
17-11-2014    101  0.2
29-11-2014    100  0.2

La columna SHIP_MODE tiene 4 valores únicos, de los cuales los primeros son:


count   pct
ship_mode                  
Standard Class  30775  60.0
Second Class    10309  20.1
First Class      7505  14.6
Same Day         2701   5.3

La columna CUSTOMER_ID tiene 1590 valores únicos, de los cuales los primeros son:


count  pct
customer_id            
PO-18850        97  0.2
BE-11335        94  0.2
JG-15805        90  0.2
SW-20755        89  0.2
MY-18295        85  0.2

La columna CUSTOMER_NAME tiene 795 valores únicos, de los cuales los primeros son:


count  pct
customer_name              
Muhammed Yedwab    108  0.2
Steven Ward        106  0.2
Gary Hwang         102  0.2
Patrick O'Brill    102  0.2
Bill Eplett        102  0.2

La columna SEGMENT tiene 3 valores únicos, de los cuales los primeros son:


count   pct
segment                 
Consumer     26518  51.7
Corporate    15429  30.1
Home Office   9343  18.2

La columna CITY tiene 3636 valores únicos, de los cuales los primeros son:


count  pct
city                     
New York City    915  1.8
Los Angeles      747  1.5
Philadelphia     537  1.0
San Francisco    510  1.0
Santo Domingo    443  0.9

La columna STATE tiene 1094 valores únicos, de los cuales los primeros son:


count  pct
state                    
California      2001  3.9
England         1499  2.9
New York        1128  2.2
Texas            985  1.9
Ile-de-France    981  1.9

La columna COUNTRY tiene 147 valores únicos, de los cuales los primeros son:


count   pct
country                   
United States   9994  19.5
Australia       2837   5.5
France          2827   5.5
Mexico          2644   5.2
Germany         2065   4.0

La columna MARKET tiene 7 valores únicos, de los cuales los primeros son:


count   pct
market             
APAC    11002  21.5
LATAM   10294  20.1
EU      10000  19.5
US       9994  19.5
EMEA     5029   9.8

La columna REGION tiene 13 valores únicos, de los cuales los primeros son:


count   pct
region              
Central  11117  21.7
South     6645  13.0
EMEA      5029   9.8
North     4785   9.3
Africa    4587   8.9

La columna PRODUCT_ID tiene 10292 valores únicos, de los cuales los primeros son:


count  pct
product_id                 
OFF-AR-10003651     35  0.1
OFF-AR-10003829     31  0.1
OFF-BI-10003708     30  0.1
OFF-BI-10002799     30  0.1
FUR-CH-10003354     28  0.1

La columna CATEGORY tiene 3 valores únicos, de los cuales los primeros son:


count   pct
category                    
Office Supplies  31273  61.0
Technology       10141  19.8
Furniture         9876  19.3

La columna SUB-CATEGORY tiene 17 valores únicos, de los cuales los primeros son:


count   pct
sub-category             
Binders        6152  12.0
Storage        5059   9.9
Art            4883   9.5
Paper          3538   6.9
Chairs         3434   6.7

La columna PRODUCT_NAME tiene 3788 valores únicos, de los cuales los primeros son:


count  pct
product_name                              
Staples                           227  0.4
Cardinal Index Tab, Clear          92  0.2
Eldon File Cart, Single Width      90  0.2
Rogers File Cart, Single Width     84  0.2
Ibico Index Tab, Clear             83  0.2

La columna QUANTITY tiene 14 valores únicos, de los cuales los primeros son:


count   pct
quantity             
2         12748  24.9
3          9682  18.9
1          8963  17.5
4          6385  12.4
5          4882   9.5

La columna ORDER_PRIORITY tiene 4 valores únicos, de los cuales los primeros son:


count   pct
order_priority             
Medium          29433  57.4
High            15501  30.2
Critical         3932   7.7
Low              2424   4.7

**Aggregation level**

First and foremost, it necessary to notice that the data at hand consists of unaggregated order records data, as there are 51290 rows with 25035 different order ids, which means there can be more than one row per order. 

To use this data to analyse by clients or to analyse by products, agggregation on the appropiate level will be necessary.

### Duplicates

**Row level**

There are 0 duplicates according to the exploration. Let us try by subsets of data:

In [72]:
subset_columns = global_superstore.columns.drop("row_id")
global_superstore.duplicated(subset=subset_columns).sum()

np.int64(0)

0 duplicates are found. Let us observe the duplicated order_ids for instance:

In [73]:
duplicated_mask = global_superstore.duplicated(subset="order_id",keep=False)
global_superstore[duplicated_mask].sort_values(by="order_id").head(4)

row_id      order_id  order_date   ship_date       ship_mode  \
34136   48312  AE-2011-9160  03-10-2011  07-10-2011  Standard Class   
29578   48313  AE-2011-9160  03-10-2011  07-10-2011  Standard Class   
5597    47298  AE-2013-1130  14-10-2013  14-10-2013        Same Day   
51156   47297  AE-2013-1130  14-10-2013  14-10-2013        Same Day   

      customer_id      customer_name   segment            city  \
34136     PO-8865  Patrick O'Donnell  Consumer           Ajman   
29578     PO-8865  Patrick O'Donnell  Consumer           Ajman   
5597      EB-4110     Eugene Barchas  Consumer  Ras al Khaymah   
51156     EB-4110     Eugene Barchas  Consumer  Ras al Khaymah   

                 state               country  postal_code market region  \
34136           'Ajman  United Arab Emirates          NaN   EMEA   EMEA   
29578           'Ajman  United Arab Emirates          NaN   EMEA   EMEA   
5597   Ra's Al Khaymah  United Arab Emirates          NaN   EMEA   EMEA   
51156  Ra's Al Khaymah  United Arab Emirates          NaN   EMEA   EMEA   

             product_id         category sub-category  \
34136  TEC-EPS-10004171       Technology     Machines   
29578  OFF-FEL-10001405  Office Supplies      Storage   
5597   FUR-BUS-10003055        Furniture    Bookcases   
51156  OFF-ACC-10004278  Office Supplies    Fasteners   

                         product_name    sales quantity  discount   profit  \
34136           Epson Calculator, Red   78.408        6       0.7  -88.992   
29578  Fellowes File Cart, Industrial   82.674        2       0.7 -157.086   
5597    Bush Stackable Bookrack, Pine  224.748        6       0.7 -232.272   
51156    Accos Paper Clips, Bulk Pack    4.248        1       0.7   -4.692   

       shipping_cost order_priority  
34136           3.87         Medium  
29578           5.69         Medium  
5597           60.08           High  
51156           0.10           High

As we suspected, order data is disaggregated per product data.

**Customer id**

Customer_ids also look duplicated, as there are more ids than there are names and exactly double, which is suspicious.

In [74]:
customer_id_name_count = global_superstore[["customer_id",
                                            "customer_name"]].value_counts().reset_index()

# min of duplicated customer name
counts_min = customer_id_name_count["customer_name"].value_counts().min()

print(f"The minimum number of appearances by customer is {counts_min}.")

# leaving this long print for explanatory purposes
customer_id_name_count.sort_values("customer_name").head(16)

The minimum number of appearances by customer is 2.


customer_id       customer_name  count
1160       AB-15       Aaron Bergman     12
19      AB-10015       Aaron Bergman     77
1199       AH-30       Aaron Hawkins     12
571     AH-10030       Aaron Hawkins     44
1154       AS-45      Aaron Smayling     13
516     AS-10045      Aaron Smayling     47
256     AB-10060     Adam Bellavance     57
1225       AB-60     Adam Bellavance     11
132     AH-10075           Adam Hart     64
883        AH-75           Adam Hart     20
1466       AS-90  Adam Shillingsburg      7
129     AS-10090  Adam Shillingsburg     64
1028      AB-105       Adrian Barton     15
162     AB-10105       Adrian Barton     62
1394      AH-120         Adrian Hane      8
533     AH-10120         Adrian Hane     46

Same customers have 2 different ids. This happens for all customers. From observation it is clear that those ids are in fact the same but with different format. Taking the example below it is obvious that it is not a case of new id formatting, just a typo. Thus we correct it by unifying ids by customer name.

In [78]:
global_superstore[global_superstore["customer_name"] == "Aaron Bergman"].sort_values(by="order_date").head(8)

# generate id mapping
id_mapping = (customer_id_name_count.sort_values("customer_id")
                    .drop_duplicates(subset="customer_name")[["customer_name","customer_id"]]
                    .set_index("customer_name")
                    .to_dict()["customer_id"])

global_superstore["customer_id"] = global_superstore["customer_name"].map(id_mapping)

**Product_id**

From the unique value counts of the above exploration, we can see that there are 10292 product ids although there only 3788 product names. Given that there are no null values for those columns, it is necessary to explore why:

In [87]:
global_superstore.groupby("product_id")["product_name"].count().describe()

count    10292.000000
mean         4.983482
std          3.410633
min          1.000000
25%          2.000000
50%          4.000000
75%          7.000000
max         35.000000
Name: product_name, dtype: float64

At least a 75% of products have 2 or more ids. Let us inspect the specific values for each.

In [92]:
product_id_name_count = global_superstore[["product_id",
                                            "product_name"]].value_counts().reset_index()

# leaving this long print for explanatory purposes
product_id_name_count.shape

(10768, 3)

There are 10768 combinations of product name and product id, which is more than there are product_ids, which means not only do products have more than 1 id, but also ids have more than 1 name.

Therefore, the approach is to create new ids for each product, grouping by their name.

In [106]:
# generate new ids
unique_product_names = global_superstore["product_name"].unique()
produc_ids = [f"PR-{id}" for id in range(0,len(unique_product_names))]

# generate new product id mapping
id_mapping = dict(zip(unique_product_names, produc_ids))

# apply changes
global_superstore["product_id"] = global_superstore["product_name"].map(id_mapping)

Now checking the change's results:

In [107]:
# checking changes
global_superstore[["product_id","product_name"]].nunique()

product_id      3788
product_name    3788
dtype: int64

#### Missing values

Drop postal_code as it adds no value due to having 80% of missing.

In [110]:
global_superstore.drop(columns="postal_code", inplace=True, errors="ignore")

#### Numerical features ranges

Sales are not integers because it actually makes reference to the product revenue, where profit would be the profit for that specific product for that order. 

Profits can be negative, which makes it very interesting to control which orders provide negative profits, as a means to minimize this loss.

Ranges for discounts are reasonable.

Shipping costs can be get very high. It is interesting to see if shipping costs are always below the sales amount:

In [112]:
(global_superstore["shipping_cost"] > global_superstore["sales"]).sum()

np.int64(0)

Then we can conclude that shipping costs are assumed by the customer.

The percentage of costs from the sale due to shipping follows the below distribution:

In [114]:
(global_superstore["shipping_cost"] / global_superstore["sales"]).describe()

count    51290.000000
mean         0.109226
std          0.069933
min          0.000000
25%          0.063136
50%          0.090254
75%          0.136364
max          0.589638
dtype: float64

#### Categorical features

#### Other checks

**Countries and regions per client**

Judging from excerpts of data, it looks like customers either order from or to many different countries.

In [76]:
global_superstore.groupby("customer_id")[["country"]].nunique().describe()

country
count  795.000000
mean    19.748428
std      3.376778
min      8.000000
25%     17.000000
50%     20.000000
75%     22.000000
max     32.000000

The above describe shows that the minimum number of countries for a customer is 8, which is not impossible but not plausible. Let us make this check with a customers' region.

In [77]:
global_superstore.groupby("customer_id")[["region"]].nunique().describe()

region
count  795.000000
mean    10.661635
std      1.164104
min      6.000000
25%     10.000000
50%     11.000000
75%     11.000000
max     13.000000

More than 75% of the customers ship to or order from at least 10 different countries, being 6 the minimum, which is very difficult to believe.

If we look at the most frequent country distribution, we see that there is no majority orders from a certain country.

In [84]:
# get most frequent country per customer 
most_frequent = global_superstore.groupby('customer_id')['country'].apply(
    lambda x: x.value_counts().idxmax()
)

# get the relative frequency for the top country
frequency = global_superstore.groupby('customer_id')['country'].apply(
    lambda x: x.value_counts().max() / x.count()
)

# get most frequent's country frequency distribution
result = pd.DataFrame({'most_frequent_country': most_frequent, 'frequency': frequency})
result.describe()

frequency
count  795.000000
mean     0.210975
std      0.072899
min      0.079365
25%      0.156250
50%      0.200000
75%      0.250000
max      0.472222

Therefore, city, state, country and region are of no further use in this project, as it is very impractical to use this data for clustering or regression.

In [85]:
global_superstore.drop(columns=["region","country","city","state"], inplace=True, errors="ignore")